### PR TITLE
qml.matrix raises an error if wire_order is not given

### PR DESCRIPTION
**Pull request by astralcai**
**Context:**
Complete the deprecation cycle for `qml.matrix()` with `wire_order=None` in certain cases.

**Description of the Change:**
`qml.matrix()` now raises an error when called on the following with no `wire_order` specified:
- tapes with more than one wire.
- quantum functions.
- QNodes if the device does not have wires specified.
- PauliWords and PauliSentences with more than one wire.

**Related Shortcut Issues:**
[sc-58182]

### CODE REVIEW

**Review by albi3ro — APPROVED** [review]

File: doc/development/deprecations.rst [hunk 1]
```diff
@@ -64,6 +64,13 @@ Pending deprecations
 Completed deprecation cycles
 ----------------------------
 
+* Calling ``qml.matrix`` without providing a ``wire_order`` on objects where the wire order could be
+  ambiguous now raises a warning. This includes tapes with multiple wires, QNodes with a device that
```
Comment (lillian542) [inline doc/development/deprecations.rst:61]:
```suggestion
  ambiguous now raises an error. This includes tapes with multiple wires, QNodes with a device that
```

**Review by lillian542 — APPROVED** [review]
One minor change in wording but looks good to me after that :)

**Comment by trbromley** [comment]
Thanks @astralcai! Note that this causes a failure in the lightning tests, e.g., https://github.com/PennyLaneAI/pennylane-lightning/actions/runs/8241518953/job/22539119250#step:14:344

CC @maliasadi @tomlqc

**Comment by astralcai** [comment]
> Thanks @astralcai! Note that this causes a failure in the lightning tests, e.g., https://github.com/PennyLaneAI/pennylane-lightning/actions/runs/8241518953/job/22539119250#step:14:344
> 
> CC @maliasadi @tomlqc

I just pushed a fix for this.

**Pull request merged**